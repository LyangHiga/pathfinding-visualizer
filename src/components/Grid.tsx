import React from "react";

import Node from "./Node";

import { GridProps } from "./types/GridTypes";

function Grid(props: GridProps) {
  const {
    grid,
    isWeighted,
    handleMouseDown,
    handleMouseEnter,
    handleMouseUp,
  } = props;

  return (
    <div style={{ textAlign: "center" }}>
      {grid.grid.map((row, rowIdx) => {
        return (
          <div key={rowIdx}>
            {row.map((node) => {
              const { row, col, val, weight } = node;
              return (
                <Node
                  key={val}
                  val={val}
                  col={col}
                  row={row}
                  onMouseDown={handleMouseDown}
                  onMouseEnter={handleMouseEnter}
                  onMouseUp={handleMouseUp}
                  weight={weight}
                  isWeighted={isWeighted}
                ></Node>
              );
            })}
          </div>
        );
      })}
    </div>
  );
}

export default Grid;
