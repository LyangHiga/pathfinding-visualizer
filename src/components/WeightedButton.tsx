import React, { useState } from "react";
import { Button, List, ListItem, ListItemText } from "@material-ui/core";
import Slider from "rc-slider";
import "rc-slider/assets/index.css";

import a from "../algorithms/a";
import styles from "../styles/NavStyles";
import WeightedButtonProps from "./types/WeightedButtonTypes";

function WeightedButton(props: WeightedButtonProps) {
  const [alpha, setAlpha] = useState(0.57);
  const [fName, setFName] = useState("A*");
  const { grid, disable, btn, handleClick, setOpenDrawer } = props;
  const classes = styles();

  const changeAlpha = (alpha: number) => setAlpha(alpha);

  const hanldeSliderChange = () => {
    if (alpha !== 0 && alpha !== 1) setFName("A*");
    else {
      if (alpha === 0) setFName("Best Fisrt Search");
      if (alpha === 1) setFName("Dijkstra");
    }
  };

  const MySlider = (
    <Slider
      defaultValue={alpha}
      min={0}
      max={1}
      onAfterChange={hanldeSliderChange}
      onChange={changeAlpha}
      step={0.01}
      disabled={disable}
      className={classes.slider}
    />
  );

  const WButton = (
    <div>
      {MySlider}
      <span className={classes.text}>Alpha: {alpha}</span>
      <Button
        className={classes.button}
        onClick={() => handleClick(a(grid, alpha))}
        disabled={disable}
      >
        {fName}
      </Button>
    </div>
  );

  const WListItem = (
    <List disablePadding>
      <ListItem divider>{MySlider}</ListItem>
      <ListItem divider className={classes.button}>
        <ListItemText className={classes.listItemText}>
          Alpha: {alpha}
        </ListItemText>
      </ListItem>
      <ListItem
        onClick={() => {
          setOpenDrawer(false);
          handleClick(a(grid, alpha));
        }}
        divider
        button
        className={classes.button}
        disabled={disable}
      >
        <ListItemText className={classes.listItemText}>{fName}</ListItemText>
      </ListItem>
    </List>
  );

  return btn ? WButton : WListItem;
}

export default WeightedButton;
