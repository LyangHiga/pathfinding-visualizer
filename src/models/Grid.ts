import Node from "./Node";
import { createAdjList, getVal } from "../helpers/gridHelper";

export default class Grid {
  public grid: Node[][];
  public start: number;
  public target: number;
  public readonly nRows: number;
  public readonly nCols: number;
  // min and max weight allowed
  public readonly min: number;
  public readonly max: number;
  constructor(
    start: number,
    target: number,
    nRows: number,
    nCols: number,
    // min and max weight
    max = 1,
    min = 1
  ) {
    this.grid = [];
    this.start = start;
    this.target = target;
    this.nCols = nCols;
    this.nRows = nRows;
    this.min = min;
    this.max = max;
    for (let row = 0; row < nRows; row++) {
      const currentRow: Node[] = [];
      for (let col = 0; col < nCols; col++) {
        // get random weight for all edges of this node
        const weight = Math.floor(Math.random() * (max - min) + min);
        const val = getVal(row, nCols, col);
        const AdjList = createAdjList(val, col, row, nRows, nCols);
        const newNode = new Node(col, row, val, start, target, AdjList, weight);
        currentRow.push(newNode);
      }
      this.grid.push(currentRow);
    }
  }
}
