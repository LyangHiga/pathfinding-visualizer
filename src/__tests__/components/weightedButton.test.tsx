// test with different g sizes, btn or drawer, disable or not
// differnt values of alpha

import React from "react";
import { shallow } from "enzyme";
import WeightedButton from "../../components/WeightedButton";
import Grid from "../../models/Grid";

const START = 10;
const TARGET = 14;

const NCOLS = 5;
const NROWS = 5;

//       0  1   2   3   4
//       5  6   7   8   9
// g =   S  11  12  13  T
//      15  16  17  18  19
//      20  21  22  23  24
describe("Node", () => {
  it("renders correctly", () => {
    const grid = new Grid(START, TARGET, NROWS, NCOLS);
    const wrapper = shallow(
      <WeightedButton
        grid={grid}
        disable={true}
        btn={true}
        handleClick={jest.fn()}
        setOpenDrawer={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
    // On the first run of this test, Jest will generate a snapshot file automatically.
  });
});

interface element {
  key: string;
  props: {
    disabled: boolean;
    defaultValue?: number;
    min?: number;
    max?: number;
    step?: number;
    children?:
      | {
          [idx: number]: number | string;
        }
      | string;
  };
}

it("Random size Grid with disable prop false and btn style, should render slider, name of the function and algorithm btn ", () => {
  // Random size grid
  const RANDOM_ROWS = Math.floor(Math.random() * 10) + 1;
  const RANDOM_COLS = Math.floor(Math.random() * 20) + 2;
  // start = 0, target = 1 are guaranteed to exist
  const g = new Grid(0, 1, RANDOM_ROWS, RANDOM_COLS);
  // disable false
  // btn ture
  const wrapper = shallow(
    <WeightedButton
      grid={g}
      disable={false}
      btn={true}
      handleClick={jest.fn()}
      setOpenDrawer={jest.fn()}
    />
  );

  const weightedBtns: element[] = wrapper.props().children;
  const slider = weightedBtns.filter((e) => e.key === "slider")[0];
  const alpha = weightedBtns.filter((e) => e.key === "alpha")[0];
  const btn = weightedBtns.filter((e) => e.key === "alg-btn")[0];

  expect(slider).not.toBeUndefined();
  expect(slider).not.toBeNull();
  expect(slider.props.disabled).toBeFalsy();
  expect(slider.props.min).toBe(0);
  expect(slider.props.max).toBe(1);
  expect(slider.props.step).toBe(0.01);

  expect(alpha).not.toBeUndefined();
  expect(alpha).not.toBeNull();
  // default val of Alpha
  expect(alpha.props.children![1]).toBe(0.57);

  expect(btn).not.toBeUndefined();
  expect(btn).not.toBeNull();
  expect(btn.props.disabled).toBeFalsy();
  // Btn name for default val of Alpha(0.57)
  expect(btn.props.children).toBe("A*");
});

it("Random size Grid with disable prop true and btn style, should render slider disabled, name of the function and algorithm btn disabled ", () => {
  // Random size grid
  const RANDOM_ROWS = Math.floor(Math.random() * 10) + 1;
  const RANDOM_COLS = Math.floor(Math.random() * 20) + 2;
  // start = 0, target = 1 are guaranteed to exist
  const g = new Grid(0, 1, RANDOM_ROWS, RANDOM_COLS);
  // disable true
  // btn true
  const wrapper = shallow(
    <WeightedButton
      grid={g}
      disable={true}
      btn={true}
      handleClick={jest.fn()}
      setOpenDrawer={jest.fn()}
    />
  );

  const weightedBtns: element[] = wrapper.props().children;
  const slider = weightedBtns.filter((e) => e.key === "slider")[0];
  const alpha = weightedBtns.filter((e) => e.key === "alpha")[0];
  const btn = weightedBtns.filter((e) => e.key === "alg-btn")[0];

  expect(slider).not.toBeUndefined();
  expect(slider).not.toBeNull();
  expect(slider.props.disabled).toBeTruthy();
  expect(slider.props.min).toBe(0);
  expect(slider.props.max).toBe(1);
  expect(slider.props.step).toBe(0.01);

  expect(alpha).not.toBeUndefined();
  expect(alpha).not.toBeNull();
  // default val of Alpha
  expect(alpha.props.children![1]).toBe(0.57);

  expect(btn).not.toBeUndefined();
  expect(btn).not.toBeNull();
  expect(btn.props.disabled).toBeTruthy();
  // Btn name for default val of Alpha(0.57)
  expect(btn.props.children).toBe("A*");
});
