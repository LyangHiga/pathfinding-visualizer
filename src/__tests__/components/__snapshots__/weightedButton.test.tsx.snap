// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node renders correctly 1`] = `
<div>
  <ComponentEnhancer(undefined)
    activeDotStyle={Object {}}
    className="makeStyles-slider-5"
    defaultValue={0.57}
    disabled={true}
    dotStyle={Object {}}
    dots={false}
    handle={[Function]}
    handleStyle={
      Array [
        Object {},
      ]
    }
    included={true}
    key="slider"
    marks={Object {}}
    max={1}
    min={0}
    onAfterChange={[Function]}
    onBeforeChange={[Function]}
    onChange={[Function]}
    prefixCls="rc-slider"
    railStyle={Object {}}
    reverse={false}
    step={0.01}
    trackStyle={
      Array [
        Object {},
      ]
    }
    vertical={false}
  />
  <span
    className="makeStyles-text-7"
    key="alpha"
  >
    Alpha: 
    0.57
  </span>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-button-4"
    disabled={true}
    key="alg-btn"
    onClick={[Function]}
  >
    A*
  </WithStyles(ForwardRef(Button))>
</div>
`;
